### PR TITLE
Adds checks to make sure json reader won't overflow

### PR DESCRIPTION
**Pull request by elstehle**
## Description
The JSON reader is currently using 32-bit offsets to index into the input's characters to lower memory footprint and for performance reasons. Hence, currently, if an input larger than `UINT_MAX` is read, the parser may return incorrect data. 

This PR adds a check that fails for inputs that could overflow. 

The longer term plan is to make the finite-state transducer stage reentrant and split up inputs larger than `UINT_MAX` into smaller chunks.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/json/nested_json_gpu.cu [hunk 1]
```diff
@@ -1076,20 +1082,25 @@ std::pair<rmm::device_uvector<PdaTokenT>, rmm::device_uvector<SymbolOffsetT>> ge
   rmm::cuda_stream_view stream,
   rmm::mr::device_memory_resource* mr)
 {
+  // Transduce() writes symbol offsets that may be as large json_in.size()-1
+  CUDF_EXPECTS(
+    json_in.size() == 0 || (json_in.size() - 1) <= std::numeric_limits<SymbolOffsetT>::max(),
+    "Given JSON input is too large");
```
Comment (vuule) [inline cpp/src/io/json/nested_json_gpu.cu:1088]: Maybe move this into a function to avoid code duplication

File: cpp/src/io/json/nested_json_gpu.cu
Comment (vuule) [inline cpp/src/io/json/nested_json_gpu.cu:1]: Could we have a test for this? What would be the peak memory use with a 2GB+1B `device_buffer_source`?

**Review by vuule — COMMENTED** [review]
Thanks for the PR! 
Posted a few suggestions.

File: cpp/src/io/json/nested_json_gpu.cu (hunk 1 above)
Comment (elstehle) [reply cpp/src/io/json/nested_json_gpu.cu]: Done 👍

File: cpp/src/io/json/nested_json_gpu.cu
Comment (elstehle) [reply cpp/src/io/json/nested_json_gpu.cu:1]: As per offline discussion, we decided that the risk of introducing a potentially flaky test that may OOM for `4 GB + 1 B` inputs is not worth it for testing such a simple check.

**Review by vuule — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**